### PR TITLE
Pass transport to Ledger signer

### LINKED ISSUE #445: Let @cosmjs/launchpad-ledger take Transport as an argument
- [x] Do the same change as it was done here: https://github.com/iov-one/iov-ledger-iovns/pull/31:
       _Give up the idea of managing Transports internally. The transport handling is too environment-specific and should be done by the caller._
- [x] Remove `optionalDependencies` and only use `@ledgerhq/hw-transport-node-hid` and `@ledgerhq/hw-transport-webusb` in demos and tests

### PR DESCRIPTION
**Pull request by willclarktech**
Closes #445

### CODE REVIEW

File: packages/launchpad-ledger/src/demo/node.ts [hunk 1]
```diff
@@ -1,6 +1,7 @@
 /* eslint-disable @typescript-eslint/naming-convention */
 import { toBase64 } from "@cosmjs/encoding";
 import { makeCosmoshubPath, makeSignDoc, StdFee, StdSignature } from "@cosmjs/launchpad";
+import LedgerTransport from "@ledgerhq/hw-transport-node-hid";
```
Comment (webmaster128) [inline packages/launchpad-ledger/src/demo/node.ts:4]: What about using the orgininal transport name form https://github.com/LedgerHQ/ledgerjs/tree/master/packages/hw-transport-node-hid/src?

File: packages/launchpad-ledger/src/demo/web.ts [hunk 2]
```diff
@@ -2,6 +2,7 @@ import { toBase64 } from "@cosmjs/encoding";
 import { AccountData, makeCosmoshubPath, StdSignDoc } from "@cosmjs/launchpad";
 import { Uint53 } from "@cosmjs/math";
 import { assert } from "@cosmjs/utils";
+import LedgerTransport from "@ledgerhq/hw-transport-webusb";
```
Comment (webmaster128) [inline packages/launchpad-ledger/src/demo/web.ts:5]: same here

File: packages/launchpad-ledger/src/launchpadledger.ts [hunk 3]
```diff
@@ -2,7 +2,7 @@ import { HdPath, Secp256k1Signature } from "@cosmjs/crypto";
 import { fromUtf8 } from "@cosmjs/encoding";
 import { makeCosmoshubPath } from "@cosmjs/launchpad";
 import { assert } from "@cosmjs/utils";
-import Transport from "@ledgerhq/hw-transport";
+import LedgerTransport from "@ledgerhq/hw-transport";
```
Comment (webmaster128) [inline packages/launchpad-ledger/src/launchpadledger.ts:5]: Why rename this interface?

File: packages/launchpad-ledger/src/launchpadledger.ts [hunk 4]
```diff
@@ -125,82 +93,20 @@ export class LaunchpadLedger {
...
-      if (trimmedErrorMessage.startsWith("No device selected")) {
-        throw new Error(
-          "You did not select a Ledger device. If you did not see your Ledger, check if the Ledger is plugged in and unlocked.",
-        );
-      }
-
-      throw error;
+    if (this.app) {
+      await this.transport.close();
+      this.app = null;
```
Comment (webmaster128) [inline packages/launchpad-ledger/src/launchpadledger.ts:109]: When the transport is provided externally, do we even want to keep this methis? We could close the transport in the caller code for symmetry.

**Review by webmaster128 — COMMENTED** [review]
Jap, much better. We now should remove the optionalDependencies and adjust the README accordingly.

File: packages/launchpad-ledger/src/launchpadledger.ts (hunk 3 above)
Comment (willclarktech) [reply packages/launchpad-ledger/src/launchpadledger.ts:5]: `Transport` is already available via the DOM lib: https://github.com/microsoft/TypeScript/blob/master/lib/lib.dom.d.ts#L20083

Comment (webmaster128) [reply packages/launchpad-ledger/src/launchpadledger.ts:5]: And we cannot override it? If I recall correctly, importing types overrides globals

Comment (willclarktech) [reply packages/launchpad-ledger/src/launchpadledger.ts:5]: We can. Do we... want to?

**Comment by willclarktech** [comment]
@webmaster128 These ledgerhq modules don't seem to install on CI. They were already failing, but because they were optional before the failure was OK: https://app.circleci.com/pipelines/github/CosmWasm/cosmjs/1369/workflows/ea95d013-ce87-41f9-a437-ae58bcecc77f/jobs/5558

Can we just skip the installation for this package since we won't be testing on CI?

**Comment by webmaster128** [comment]
> Can we just skip the installation for this package since we won't be testing on CI?

I like the idea. Do you have an idea how to skip the installation of a `devDependency`?

File: packages/launchpad-ledger/src/launchpadledger.ts (hunk 3 above)
Comment (webmaster128) [reply packages/launchpad-ledger/src/launchpadledger.ts:5]: I think we should. This is only a type and not a global object. `Transport` is undefined in JavaScript (node.js, Firefox and Chrome)

**Comment by willclarktech** [comment]
@webmaster128 I thought I did: https://github.com/CosmWasm/cosmjs/pull/446/commits/27e0273390956e4362b64c7d6c48214f99f959fd
But it fails both locally and on CI (but for different reasons).

File: .circleci/config.yml [hunk 5]
```diff
@@ -41,6 +41,9 @@ jobs:
           name: Restore Yarn Package Cache
           keys:
             - yarn-packages-{{ checksum "yarn.lock" }}
+      - run:
+        name: Install libusb
+        command: sudo apt-get install libusb
```
Comment (webmaster128) [inline .circleci/config.yml:46]: you probably need someting like libusb-dev to compile against libusb headers

File: packages/launchpad-ledger/src/ledgersigner.spec.ts [hunk 6]
```diff
@@ -12,12 +12,29 @@ import {
   StdFee,
 } from "@cosmjs/launchpad";
 import { assert, sleep } from "@cosmjs/utils";
+import LedgerTransport from "@ledgerhq/hw-transport";
```
Comment (webmaster128) [inline packages/launchpad-ledger/src/ledgersigner.spec.ts:15]:
```suggestion
import Transport from "@ledgerhq/hw-transport";
```

File: packages/launchpad-ledger/src/ledgersigner.spec.ts [hunk 7]
```diff
@@ -12,12 +12,29 @@ import {
...
 
 import { LedgerSigner } from "./ledgersigner";
 import { pendingWithoutLedger, pendingWithoutWasmd, wasmd } from "./testutils.spec";
 
 const interactiveTimeout = 120_000;
 
+async function createLedgerTransport(): Promise<LedgerTransport> {
```
Comment (webmaster128) [inline packages/launchpad-ledger/src/ledgersigner.spec.ts:22]:
```suggestion
async function createLedgerTransport(): Promise<Transport> {
```

File: packages/launchpad-ledger/src/ledgersigner.spec.ts [hunk 8]
```diff
@@ -28,11 +45,20 @@ describe("LedgerSigner", () => {
   const defaultSequence = "0";
   const defaultAccountNumber = "42";
   const defaultRecipient = "cosmos1p6xs63q4g7np99ttv5nd3yzkt8n4qxa47w8aea";
+  let ledgerTransport: LedgerTransport;
```
Comment (webmaster128) [inline packages/launchpad-ledger/src/ledgersigner.spec.ts:48]:
```suggestion
  let ledgerTransport: Transport;
```

File: packages/launchpad-ledger/src/ledgersigner.ts [hunk 9]
```diff
@@ -7,6 +7,7 @@ import {
   StdSignDoc,
 } from "@cosmjs/launchpad";
 import { serializeSignDoc, SignResponse } from "@cosmjs/launchpad";
+import LedgerTransport from "@ledgerhq/hw-transport";
```
Comment (webmaster128) [inline packages/launchpad-ledger/src/ledgersigner.ts:10]:
```suggestion
import Transport from "@ledgerhq/hw-transport";
```

File: packages/launchpad-ledger/src/ledgersigner.ts [hunk 10]
```diff
@@ -15,9 +16,9 @@ export class LedgerSigner implements OfflineSigner {
   private readonly hdPaths: readonly HdPath[];
   private accounts?: readonly AccountData[];
 
-  public constructor(options: LaunchpadLedgerOptions = {}) {
+  public constructor(transport: LedgerTransport, options: LaunchpadLedgerOptions = {}) {
```
Comment (webmaster128) [inline packages/launchpad-ledger/src/ledgersigner.ts:19]:
```suggestion
  public constructor(transport: Transport, options: LaunchpadLedgerOptions = {}) {
```

**Review by webmaster128 — APPROVED** [review]

**Pull request merged**